src/game/Unit.cpp

```diff
@@ -2201,132 +2201,110 @@ MeleeHitOutcome Unit::RollMeleeOutcomeAgainst(const Unit* pVictim, WeaponAttackT
-                        return MELEE_HIT_BLOCK_CRIT;
-                    }
-                }
-                DEBUG_FILTER_LOG(LOG_FILTER_COMBAT, "RollMeleeOutcomeAgainst: BLOCK <%d, %d)", sum - tmp, sum);
+                DEBUG_LOG("RollMeleeOutcomeAgainst: BLOCKED CRIT");
+                return MELEE_HIT_BLOCK_CRIT;
+            }
+            else
+            {
+                DEBUG_FILTER_LOG(LOG_FILTER_COMBAT, "RollMeleeOutcomeAgainst: BLOCK %d)", tmp);
                 return MELEE_HIT_BLOCK;
             }
         }
     }
 
-    // Critical chance
-    tmp = crit_chance;
-
-    if (tmp > 0 && roll < (sum += tmp))
+    tmp = crit_chance + skillBonus;
+    if (roll < (sum += tmp))
     {
-        DEBUG_FILTER_LOG(LOG_FILTER_COMBAT, "RollMeleeOutcomeAgainst: CRIT <%d, %d)", sum - tmp, sum);
+        DEBUG_FILTER_LOG(LOG_FILTER_COMBAT, "RollMeleeOutcomeAgainst: CRIT %d)", crit_chance);
         return MELEE_HIT_CRIT;
     }
 
-    tmp = (victimDefenseSkill < victimMaxSkillValueForLevel) ? victimDefenseSkill : victimMaxSkillValueForLevel;
+    // Max 40% chance to score a glancing blow against mobs that are higher level
+    // (can do only players and pets and not with ranged weapon)
+    if ((GetTypeId() == TYPEID_PLAYER || ((Creature*)this)->IsPet())
+        && pVictim->GetTypeId() != TYPEID_PLAYER && !((Creature*)pVictim)->IsPet()
+        && getLevel() < pVictim->getLevel() && attType != RANGED_ATTACK && !SpellCasted)
+    {
+        roll = urand(0, 10000);
+        tmp = 1000;
+
+        if (attackerWeaponSkill < attackerMaxSkillValueForLevel)
+            tmp += (victimDefenseSkill - attackerWeaponSkill) * 200;
+        else
+            tmp += (victimDefenseSkill - attackerMaxSkillValueForLevel) * 200;
 
+        if (roll < (tmp = (tmp > 4000) ? 4000 : tmp))
+        {
+            DEBUG_FILTER_LOG(LOG_FILTER_COMBAT, "RollMeleeOutcomeAgainst: GLANCING %d)", tmp);
+            return MELEE_HIT_GLANCING;
+        }
+    }
     // mobs can score crushing blows if they're 3 or more levels above victim
     // having defense above your maximum (from items, talents etc.) has no effect
     // mob's level * 5 - player's current defense skill - add 2% chance per lacking skill point, min. is 15%
-    if ((getLevel() - 3) >= pVictim->getLevel() && !SpellCasted
-        && roll < (tmp = (((attackerMaxSkillValueForLevel - tmp) * 200) - 1500)))
+    else if ((getLevel() - 3) >= pVictim->getLevel() && !SpellCasted)
     {
+        roll = urand(0, 100) * 100;
         uint32 typeId = GetTypeId();
-        if ((typeId == TYPEID_UNIT && !(GetOwnerGuid() && GetOwner()->GetTypeId() == TYPEID_PLAYER)
+        tmp = (victimDefenseSkill < victimMaxSkillValueForLevel) ? victimDefenseSkill : victimMaxSkillValueForLevel;
+
+        if (roll < (tmp = (((attackerMaxSkillValueForLevel - tmp) * 200) - 1500))
+            && ((typeId == TYPEID_UNIT && !(GetOwnerGuid() && GetOwner()->GetTypeId() == TYPEID_PLAYER)
```

Comment:
probably move much of these 'conditioned' checks up to where they belong, with the other conditions up in the 'else if' at line 2298, and actually only keep the roll check here